pkg/daemon/file_writers.go

```diff
@@ -0,0 +1,302 @@
+		if err != nil {
+			return fmt.Errorf("could not decode file %q: %w", file.Path, err)
+		}
+
+		mode := defaultFilePermissions
+		if file.Mode != nil {
+			mode = os.FileMode(*file.Mode)
+		}
+
+		// set chown if file information is provided
+		uid, gid, err := getFileOwnership(file)
+		if err != nil {
+			return fmt.Errorf("failed to retrieve file ownership for file %q: %w", file.Path, err)
+		}
+		if err := createOrigFile(file.Path, file.Path); err != nil {
+			return err
+		}
+		if err := writeFileAtomically(file.Path, decodedContents, defaultDirectoryPermissions, mode, uid, gid); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// writeUnit writes s systemd unit and its dropins to disk
+func writeUnit(u ign3types.Unit, systemdRoot string, isCoreOSVariant bool) error {
+	if err := writeDropins(u, systemdRoot, isCoreOSVariant); err != nil {
+		return err
+	}
+
+	// write (or cleanup) path in /etc/systemd/system
+	fpath := filepath.Join(systemdRoot, u.Name)
+	if u.Mask != nil && *u.Mask {
+		// if the unit is masked, symlink fpath to /dev/null and return early.
+
+		glog.V(2).Info("Systemd unit masked")
+		if err := os.RemoveAll(fpath); err != nil {
+			return fmt.Errorf("failed to remove unit %q: %w", u.Name, err)
+		}
+		glog.V(2).Infof("Removed unit %q", u.Name)
+
+		if err := renameio.Symlink(pathDevNull, fpath); err != nil {
+			return fmt.Errorf("failed to symlink unit %q to %s: %w", u.Name, pathDevNull, err)
+		}
+		glog.V(2).Infof("Created symlink unit %q to %s", u.Name, pathDevNull)
+
+		// Return early since we don't need to write the file contents in this case.
+		return nil
+	}
+
+	if u.Contents != nil && *u.Contents != "" {
+		glog.Infof("Writing systemd unit %q", u.Name)
+		if _, err := os.Stat(withUsrPath(fpath)); err == nil &&
+			isCoreOSVariant {
+			if err := createOrigFile(withUsrPath(fpath), fpath); err != nil {
+				return err
+			}
+		}
+		if err := writeFileAtomicallyWithDefaults(fpath, []byte(*u.Contents)); err != nil {
+			return fmt.Errorf("failed to write systemd unit %q: %w", u.Name, err)
+		}
+
+		glog.V(2).Infof("Successfully wrote systemd unit %q: ", u.Name)
+	} else if u.Mask != nil && !*u.Mask {
+		// if mask is explicitly set to false, make sure to remove a previous mask
+		// see https://bugzilla.redhat.com/show_bug.cgi?id=1966445
+		// Note that this does not catch all cleanup cases; for example, if the previous machine config specified
+		// Contents, and the current one does not, the previous content will not get cleaned up. For now we're ignoring some
+		// of those edge cases rather than introducing more complexity.
+		glog.V(2).Infof("Ensuring systemd unit %q has no mask at %q", u.Name, fpath)
+		if err := os.RemoveAll(fpath); err != nil {
+			return fmt.Errorf("failed to cleanup %s: %w", fpath, err)
+		}
+	}
+
+	return nil
+}
+
+// writeUnit writes systemd units and their dropins to disk
```

Comment:
nit: the function name is `writeUnits` with an s

Reply:
Done!